http-body-util/src/combinators/with_trailers.rs

```diff
@@ -0,0 +1,158 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures_util::ready;
+use http::HeaderMap;
+use http_body::{Body, Frame};
+use pin_project_lite::pin_project;
+
+pin_project! {
+    /// Adds trailers to a body.
+    ///
+    /// See [`BodyExt::with_trailers`] for more details.
+    pub struct WithTrailers<T, F> {
+        #[pin]
+        state: State<T, F>,
+    }
+}
+
+impl<T, F> WithTrailers<T, F> {
+    pub(crate) fn new(body: T, trailers: F) -> Self {
+        Self {
+            state: State::PollBody {
+                body,
+                trailers: Some(trailers),
+            },
+        }
+    }
+}
+
+pin_project! {
+    #[project = StateProj]
+    enum State<T, F> {
+        PollBody {
+            #[pin]
+            body: T,
+            trailers: Option<F>,
+        },
+        PollTrailers {
+            #[pin]
+            trailers: F,
+        },
+        Trailers {
+            trailers: Option<HeaderMap>,
+        }
+    }
+}
+
+impl<T, F> Body for WithTrailers<T, F>
+where
+    T: Body,
+    F: Future<Output = Option<Result<HeaderMap, T::Error>>>,
+{
+    type Data = T::Data;
+    type Error = T::Error;
+
+    fn poll_frame(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+        loop {
+            let mut this = self.as_mut().project();
+
+            let new_state: State<_, _> = match this.state.as_mut().project() {
+                StateProj::PollBody { body, trailers } => match ready!(body.poll_frame(cx)?) {
+                    Some(frame) => {
+                        return Poll::Ready(Some(Ok(frame)));
+                    }
```

Comment:
IIUC, if `frame` is a trailer here, the `WithTrailer` would yield multiple trailer frames. Is that intentional? Even if so, I think it would be nice to document that behavior.

Reply:
Returning multiple trailers frames is likely to cause problems for consumers. Also, I was wondering when I saw this PR, should this merge trailers? I think probably it should, such as `an_h2_body.with_trailers(t)` or `body.with_trailers(t1).with_trailers(t2)`.

Reply:
Fixed! It'll now merge trailers returned by the inner body :)